src/System.Text.Json/ref/System.Text.Json.cs

```diff
@@ -156,6 +245,113 @@ public abstract partial class JsonNamingPolicy
+    }
+    public partial class JsonObject : System.Text.Json.JsonNode, System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, System.Text.Json.JsonNode>>, System.Collections.IEnumerable
+    {
+        public JsonObject() { }
+        public JsonObject(System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, System.Text.Json.JsonNode>> jsonProperties) { }
+        public System.Text.Json.JsonNode this[string key] { get { throw null; } set { } }
+        public System.Collections.Generic.ICollection<string> PropertyNames { get { throw null; } }
+        public System.Collections.Generic.ICollection<System.Text.Json.JsonNode> Values { get { throw null; } }
+        public void Add(System.Collections.Generic.KeyValuePair<string, System.Text.Json.JsonNode> jsonProperty) { }
+        public void Add(string propertyName, bool propertyValue) { }
+        public void Add(string propertyName, byte propertyValue) { }
+        public void Add(string propertyName, System.Collections.Generic.IEnumerable<System.Text.Json.JsonNode> propertyValue) { }
+        public void Add(string propertyName, double propertyValue) { }
+        public void Add(string propertyName, short propertyValue) { }
+        public void Add(string propertyName, int propertyValue) { }
+        public void Add(string propertyName, long propertyValue) { }
+        [System.CLSCompliantAttribute(false)]
+        public void Add(string propertyName, sbyte propertyValue) { }
+        public void Add(string propertyName, float propertyValue) { }
+        public void Add(string propertyName, string propertyValue) { }
+        public void Add(string propertyName, System.Text.Json.JsonArray propertyValue) { }
+        public void Add(string propertyName, System.Text.Json.JsonNode propertyValue) { }
+        [System.CLSCompliantAttribute(false)]
+        public void Add(string propertyName, ushort propertyValue) { }
+        [System.CLSCompliantAttribute(false)]
+        public void Add(string propertyName, uint propertyValue) { }
+        [System.CLSCompliantAttribute(false)]
+        public void Add(string propertyName, ulong propertyValue) { }
+        public void AddRange(System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, System.Text.Json.JsonNode>> jsonProperties) { }
+        public bool ContainsProperty(string propertyName) { throw null; }
+        public System.Collections.Generic.IEnumerable<System.Text.Json.JsonNode> GetAllPrimaryTypedValues() { throw null; }
+        public System.Collections.Generic.IEnumerable<System.Text.Json.JsonNode> GetAllValuesByPropertyName(string propertyName) { throw null; }
+        public System.Collections.Generic.IEnumerator<System.Collections.Generic.KeyValuePair<string, System.Text.Json.JsonNode>> GetEnumerator() { throw null; }
+        public System.Text.Json.JsonObject GetProperty(string propertyName) { throw null; }
```

Comment:
For the next three methods, how do we plan to deal with duplicates?

Reply:
`JsonObject` can't have different properties with the same name.

Reply:
> JsonObject can't have different properties with the same name.

JSON objects can... there's no uniqueness constraint `{ "prop": true, "prop": false }` is perfectly valid (but confusing) JSON.

Reply:
Here it throws an error "Duplicate key": https://jsonlint.com/ (but maybe it's the constraint of this specific validator, not Json itself).

Reply:
That's helping you to write "good JSON".  But "legal JSON" allows duplicates.  The spec is really tiny: https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf

> **6 Objects**
> An object structure is represented as a pair of curly bracket tokens surrounding zero or more name/value pairs. A  name  is  a string.  A  single  colon  token  follows  each  name,  separating  the  name from  the value.  A  single comma token separates a value from a following name. The JSON syntax does not impose any restrictions on the strings used  as  names,  **does  not  require  that  name strings be  unique**,  and  does  not  assign  any significance to the ordering of name/value pairs. These are all semantic considerations that may be defined by JSON processors or in specifications defining specific uses of JSON for data interchange.

<sub>(emphasis mine)</sub>

Reply:
I see, thank you! I'll take a look at this.

Reply:
I can change `GetProperty` to return `IEnumerable<JsonNode>` and name it `GetAllProperties`. I think it might be valuable to also add `GetJsonObject(string propertyName)` returning `JsonObject` (and the same for `JsonArray`) to allow quicker access to nested members if user knows how the structure of Json look like (without all of casting and checks). What do you think?

Reply:
(Try)GetProperty(string) => JsonNode and GetAllProperties(string) => `IEnumerable<JsonNode>` seem compatible with each other as long as you think someone can reason about both existing (the simple one being "get first" (or last, whatever)).

(Try)GetObjectProperty(string) and (Try)GetArrayProperty(string) also seem reasonable, as accelerators for GetAllProperties(string).FirstOrDefault(p => p.ValueType == JsonValueType.Array); (or Object, or Last).

Reply:
I added all of those methods, I'm thinking of changing indexer? I left it to behave as `GetProperty` (return `JsonNode` instead of `IEnumerable<JsonNode>`), because I think in most cases property names will be unique (that's why I also left `GetProperty`)?

Reply:
The indexer returning a JsonNode sounds reasonable.  I agree that the keys are usually unique.